File: backend/src/api/v1/mcp.py
"""
MCP API Endpoints

Provides REST API endpoints for managing MCP servers and invoking tools.
"""

from typing import Any, Dict, List, Optional

from fastapi import APIRouter, HTTPException, status
from pydantic import BaseModel, Field

from src.mcp import (
    get_mcp_bus,
    MCPServerConfig,
    MCPTransport,
    ToolDefinition,
    ServerInfo,
)
from src.mcp.bus import ServerNotFoundError, ToolInvocationError
from src.mcp.registry import ToolNotFoundError

router = APIRouter()


# ==================== Request/Response Schemas ====================


class ServerRegisterRequest(BaseModel):
    """Request to register an MCP server."""
    name: str = Field(..., description="Unique name for the server")
    transport: str = Field(default="stdio", description="Transport type (stdio, sse)")
    command: Optional[str] = Field(None, description="Command to run (for STDIO)")
    args: List[str] = Field(default_factory=list, description="Command arguments")
    env: Dict[str, str] = Field(default_factory=dict, description="Environment variables")
    url: Optional[str] = Field(None, description="Server URL (for SSE)")
    timeout: float = Field(default=60.0, description="Timeout in seconds")


class ToolInvokeRequest(BaseModel):
    """Request to invoke an MCP tool."""
    tool_path: str = Field(..., description="Full tool path (server.tool_name)")
    arguments: Dict[str, Any] = Field(default_factory=dict, description="Tool arguments")
    timeout: Optional[float] = Field(None, description="Optional timeout override")


class ToolInvokeResponse(BaseModel):
    """Response from tool invocation."""
    status: str
    tool_path: str
    result: Optional[Any] = None
    error: Optional[str] = None
    execution_time: float


class ServerStatusResponse(BaseModel):
    """Server status response."""
    name: str
    status: str
    connected: bool
    tool_count: int
    connected_at: Optional[str] = None
    error: Optional[str] = None


class BusHealthResponse(BaseModel):
    """MCP Bus health response."""
    bus_running: bool
    total_servers: int
    connected_servers: int
    total_tools: int
    servers: Dict[str, Any]


# ==================== API Endpoints ====================


@router.get("/health", response_model=BusHealthResponse)
async def get_mcp_health():
    """Get MCP Bus health status."""
    bus = get_mcp_bus()
    health = await bus.health_check()
    return health


@router.get("/servers", response_model=List[ServerStatusResponse])
async def list_servers():
    """List all registered MCP servers."""
    bus = get_mcp_bus()
    servers = await bus.get_all_servers()

    return [
        ServerStatusResponse(
            name=s.name,
            status=s.status,
            connected=s.status == "connected",
            tool_count=s.tool_count,
            connected_at=s.connected_at.isoformat() if s.connected_at else None,
            error=s.error_message
        )
        for s in servers
    ]


@router.post("/servers", response_model=ServerStatusResponse, status_code=status.HTTP_201_CREATED)
async def register_server(request: ServerRegisterRequest):
    """Register a new MCP server."""
    bus = get_mcp_bus()

    try:
        transport = MCPTransport(request.transport)
    except ValueError:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Invalid transport type: {request.transport}. Valid: stdio, sse, websocket"
        )

    config = MCPServerConfig(
        name=request.name,
        transport=transport,
        command=request.command,
        args=request.args,
        env=request.env,
        url=request.url,
        timeout=request.timeout
    )

    try:
        server_info = await bus.register_server(request.name, config)
        return ServerStatusResponse(
            name=server_info.name,
            status=server_info.status,
            connected=server_info.status == "connected",
            tool_count=server_info.tool_count,
            connected_at=server_info.connected_at.isoformat() if server_info.connected_at else None,
            error=server_info.error_message
        )
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )


@router.get("/servers/{server_name}", response_model=ServerStatusResponse)
async def get_server(server_name: str):
    """Get status of a specific MCP server."""
    bus = get_mcp_bus()

    try:
        server_info = await bus.get_server_status(server_name)
        return ServerStatusResponse(
            name=server_info.name,
            status=server_info.status,
            connected=server_info.status == "connected",
            tool_count=server_info.tool_count,
            connected_at=server_info.connected_at.isoformat() if server_info.connected_at else None,
            error=server_info.error_message
        )
    except ServerNotFoundError:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Server not found: {server_name}"
        )


@router.delete("/servers/{server_name}", status_code=status.HTTP_204_NO_CONTENT)
async def unregister_server(server_name: str):
    """Unregister an MCP server."""
    bus = get_mcp_bus()

    try:
        await bus.unregister_server(server_name)
    except ServerNotFoundError:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Server not found: {server_name}"
        )


@router.post("/servers/{server_name}/reconnect", response_model=ServerStatusResponse)
async def reconnect_server(server_name: str):
    """Reconnect to an MCP server."""
    bus = get_mcp_bus()

    try:
        await bus.reconnect_server(server_name)
        server_info = await bus.get_server_status(server_name)
        return ServerStatusResponse(
            name=server_info.name,
            status=server_info.status,
            connected=server_info.status == "connected",
            tool_count=server_info.tool_count,
            connected_at=server_info.connected_at.isoformat() if server_info.connected_at else None,
            error=server_info.error_message
        )
    except ServerNotFoundError:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Server not found: {server_name}"
        )


@router.get("/tools")
async def list_tools(server_name: Optional[str] = None):
    """List all available MCP tools."""
    bus = get_mcp_bus()
    tools = await bus.get_available_tools(server_name)

    return {
        "total": len(tools),
        "tools": [
            {
                "name": t.name,
                "description": t.description,
                "server": t.server_name,
                "input_schema": t.input_schema
            }
            for t in tools
        ]
    }


@router.get("/tools/{tool_path:path}")
async def get_tool(tool_path: str):
    """Get details of a specific MCP tool."""
    bus = get_mcp_bus()

    try:
        tool = await bus.registry.get_tool(tool_path)
        return {
            "name": tool.name,
            "description": tool.description,
            "server": tool.server_name,
            "input_schema": tool.input_schema
        }
    except ToolNotFoundError:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Tool not found: {tool_path}"
        )


@router.post("/tools/invoke", response_model=ToolInvokeResponse)
async def invoke_tool(request: ToolInvokeRequest):
    """Invoke an MCP tool."""
    bus = get_mcp_bus()

    try:
        result = await bus.invoke_tool(
            request.tool_path,
            request.arguments,
            request.timeout
        )

        return ToolInvokeResponse(
            status=result.status,
            tool_path=result.tool_path,
            result=result.result,
            error=result.error,
            execution_time=result.execution_time
        )
    except ToolNotFoundError:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Tool not found: {request.tool_path}"
        )
    except ServerNotFoundError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )
    except ToolInvocationError as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e)
        )


@router.post("/servers/{server_name}/tools")
async def register_tool(server_name: str, tool: Dict[str, Any]):
    """Manually register a tool for a server."""
    bus = get_mcp_bus()

    try:
        tool_def = ToolDefinition(
            name=tool.get("name", ""),
            description=tool.get("description", ""),
            input_schema=tool.get("input_schema", {}),
            server_name=server_name
        )
        await bus.register_tool_manually(server_name, tool_def)
        return {"status": "registered", "tool": tool_def.name}
    except ServerNotFoundError:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Server not found: {server_name}"
        )
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )

File: frontend-v2/src/services/mcp.ts
/**
 * MCP (Model Context Protocol) API Service
 *
 * Handles MCP server management, tool discovery, and tool invocation.
 */

import { get, post, del } from './api';
import type {
  MCPServerRegisterRequest,
  MCPServerStatusResponse,
  MCPBusHealthResponse,
  MCPToolDefinition,
  MCPToolListResponse,
  MCPToolInvokeRequest,
  MCPToolInvokeResponse,
} from '../types/api';

// =============================================================================
// API Endpoints
// =============================================================================

const MCP_ENDPOINTS = {
  HEALTH: '/mcp/health',
  SERVERS: '/mcp/servers',
  SERVER_BY_NAME: (name: string) => `/mcp/servers/${name}`,
  SERVER_RECONNECT: (name: string) => `/mcp/servers/${name}/reconnect`,
  SERVER_TOOLS: (name: string) => `/mcp/servers/${name}/tools`,
  TOOLS: '/mcp/tools',
  TOOL_BY_PATH: (path: string) => `/mcp/tools/${path}`,
  INVOKE: '/mcp/tools/invoke',
} as const;

// =============================================================================
// MCP Service
// =============================================================================

/**
 * MCP service class for managing MCP servers and tools
 */
class MCPService {
  // =========================================================================
  // Health & Status
  // =========================================================================

  /**
   * Get MCP Bus health status
   *
   * Returns overall health including connected servers and available tools.
   *
   * @returns MCP Bus health information
   */
  async getMCPHealth(): Promise<MCPBusHealthResponse> {
    return get<MCPBusHealthResponse>(MCP_ENDPOINTS.HEALTH);
  }

  // =========================================================================
  // Server Management
  // =========================================================================

  /**
   * Get all registered MCP servers
   *
   * @returns List of server status responses
   */
  async getServers(): Promise<MCPServerStatusResponse[]> {
    return get<MCPServerStatusResponse[]>(MCP_ENDPOINTS.SERVERS);
  }

  /**
   * Get status of a specific MCP server
   *
   * @param name - Server name
   * @returns Server status
   */
  async getServer(name: string): Promise<MCPServerStatusResponse> {
    return get<MCPServerStatusResponse>(MCP_ENDPOINTS.SERVER_BY_NAME(name));
  }

  /**
   * Register a new MCP server
   *
   * @param data - Server configuration
   * @returns Registered server status
   */
  async registerServer(data: MCPServerRegisterRequest): Promise<MCPServerStatusResponse> {
    return post<MCPServerStatusResponse>(MCP_ENDPOINTS.SERVERS, data);
  }

  /**
   * Unregister an MCP server
   *
   * @param name - Server name to unregister
   */
  async unregisterServer(name: string): Promise<void> {
    await del(MCP_ENDPOINTS.SERVER_BY_NAME(name));
  }

  /**
   * Reconnect to an MCP server
   *
   * Useful when a server connection has been lost.
   *
   * @param name - Server name to reconnect
   * @returns Updated server status
   */
  async reconnectServer(name: string): Promise<MCPServerStatusResponse> {
    return post<MCPServerStatusResponse>(MCP_ENDPOINTS.SERVER_RECONNECT(name));
  }

  // =========================================================================
  // Tool Management
  // =========================================================================

  /**
   * Get all available MCP tools
   *
   * @param serverName - Optional server name to filter tools
   * @returns List of available tools
   */
  async getTools(serverName?: string): Promise<MCPToolListResponse> {
    const params: Record<string, unknown> = {};
    if (serverName) {
      params.server_name = serverName;
    }
    return get<MCPToolListResponse>(MCP_ENDPOINTS.TOOLS, params);
  }

  /**
   * Get details of a specific tool
   *
   * @param toolPath - Full tool path (server.tool_name)
   * @returns Tool definition
   */
  async getTool(toolPath: string): Promise<MCPToolDefinition> {
    return get<MCPToolDefinition>(MCP_ENDPOINTS.TOOL_BY_PATH(toolPath));
  }

  /**
   * Invoke an MCP tool
   *
   * @param toolPath - Full tool path (server.tool_name)
   * @param args - Tool arguments
   * @param timeout - Optional timeout override in seconds
   * @returns Tool invocation result
   */
  async invokeTool(
    toolPath: string,
    args: Record<string, unknown> = {},
    timeout?: number
  ): Promise<MCPToolInvokeResponse> {
    const data: MCPToolInvokeRequest = {
      tool_path: toolPath,
      arguments: args,
      timeout,
    };
    return post<MCPToolInvokeResponse>(MCP_ENDPOINTS.INVOKE, data);
  }

  /**
   * Manually register a tool for a server
   *
   * This is useful for adding tools that weren't auto-discovered.
   *
   * @param serverName - Server to register the tool for
   * @param tool - Tool definition
   */
  async registerTool(
    serverName: string,
    tool: Omit<MCPToolDefinition, 'server'>
  ): Promise<{ status: string; tool: string }> {
    return post<{ status: string; tool: string }>(
      MCP_ENDPOINTS.SERVER_TOOLS(serverName),
      tool
    );
  }

  // =========================================================================
  // Convenience Methods
  // =========================================================================

  /**
   * Check if MCP Bus is healthy
   */
  async isHealthy(): Promise<boolean> {
    try {
      const health = await this.getMCPHealth();
      return health.bus_running;
    } catch {
      return false;
    }
  }

  /**
   * Get connected servers only
   */
  async getConnectedServers(): Promise<MCPServerStatusResponse[]> {
    const servers = await this.getServers();
    return servers.filter((s) => s.connected);
  }

  /**
   * Get disconnected servers
   */
  async getDisconnectedServers(): Promise<MCPServerStatusResponse[]> {
    const servers = await this.getServers();
    return servers.filter((s) => !s.connected);
  }

  /**
   * Get tools for a specific server
   */
  async getServerTools(serverName: string): Promise<MCPToolDefinition[]> {
    const response = await this.getTools(serverName);
    return response.tools;
  }

  /**
   * Register a STDIO-based MCP server
   */
  async registerStdioServer(
    name: string,
    command: string,
    args: string[] = [],
    env: Record<string, string> = {},
    timeout = 60
  ): Promise<MCPServerStatusResponse> {
    return this.registerServer({
      name,
      transport: 'stdio',
      command,
      args,
      env,
      timeout,
    });
  }

  /**
   * Register an SSE-based MCP server
   */
  async registerSseServer(
    name: string,
    url: string,
    timeout = 60
  ): Promise<MCPServerStatusResponse> {
    return this.registerServer({
      name,
      transport: 'sse',
      url,
      timeout,
    });
  }

  /**
   * Quick invoke with just tool path and args
   */
  async quickInvoke(
    toolPath: string,
    args: Record<string, unknown>
  ): Promise<unknown> {
    const result = await this.invokeTool(toolPath, args);
    if (result.status === 'error') {
      throw new Error(result.error || 'Tool invocation failed');
    }
    return result.result;
  }

  /**
   * Search tools by name or description
   */
  async searchTools(query: string): Promise<MCPToolDefinition[]> {
    const response = await this.getTools();
    const lowerQuery = query.toLowerCase();
    return response.tools.filter(
      (t) =>
        t.name.toLowerCase().includes(lowerQuery) ||
        t.description.toLowerCase().includes(lowerQuery)
    );
  }

  /**
   * Get a summary of all MCP resources
   */
  async getSummary(): Promise<{
    health: MCPBusHealthResponse;
    servers: MCPServerStatusResponse[];
    tools: MCPToolListResponse;
  }> {
    const [health, servers, tools] = await Promise.all([
      this.getMCPHealth(),
      this.getServers(),
      this.getTools(),
    ]);
    return { health, servers, tools };
  }
}

// =============================================================================
// Export Singleton Instance
// =============================================================================

export const mcpService = new MCPService();

// Export individual functions for convenience
export const getMCPHealth = mcpService.getMCPHealth.bind(mcpService);
export const getServers = mcpService.getServers.bind(mcpService);
export const getServer = mcpService.getServer.bind(mcpService);
export const registerServer = mcpService.registerServer.bind(mcpService);
export const unregisterServer = mcpService.unregisterServer.bind(mcpService);
export const reconnectServer = mcpService.reconnectServer.bind(mcpService);
export const getTools = mcpService.getTools.bind(mcpService);
export const getTool = mcpService.getTool.bind(mcpService);
export const invokeTool = mcpService.invokeTool.bind(mcpService);

export default mcpService;
